File: app/src/main/java/com/github/coderbuck/covid19/CovidApiUtils.java
package com.github.coderbuck.covid19;

import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import retrofit2.Call;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class CovidApiUtils {

    private static OkHttpClient okHttpClient = new OkHttpClient.Builder()
            .callTimeout(1, TimeUnit.MINUTES)
            .connectTimeout(1, TimeUnit.MINUTES)
            .readTimeout(1, TimeUnit.MINUTES)
            .build();

    private static final Retrofit retrofit = new Retrofit.Builder()
            .client(okHttpClient)
            .baseUrl(CovidApi.BASE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build();

    private static CovidApi covidApi;

    public static CovidApi getApi() {
        if (covidApi == null) {
            covidApi = retrofit.create(CovidApi.class);
        }
        return covidApi;
    }


    public static Call<OverAll> getOverAllHistory() {
        return getApi().getOverAllHistory();
    }

    public static Call<OverAll> getOverAllLatest() {
        return getApi().getOverAllLatest();
    }

    public static Call<AreaData> getAreaDataLatest() {
        return getApi().getAreaDataLatest();
    }

    public static Call<AreaData> getAreaDataHistory() {
        return getApi().getAreaDataHistory();
    }
}

File: app/src/main/java/com/github/coderbuck/covid19/CovidApiUtils.kt
package com.github.coderbuck.covid19

import com.github.coderbuck.covid19.bean.Area
import com.github.coderbuck.covid19.bean.OverAll
import com.github.coderbuck.covid19.bean.Result
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Call
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

object CovidApiUtils {
    private val okHttpClient = OkHttpClient.Builder()
            .callTimeout(1, TimeUnit.MINUTES)
            .connectTimeout(1, TimeUnit.MINUTES)
            .readTimeout(1, TimeUnit.MINUTES)
            .addInterceptor(UserAgentInterceptor("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.149 Safari/537.36"))
            .addInterceptor(createHttpLoggingInterceptor())
            .build()

    private fun createHttpLoggingInterceptor(): Interceptor {
        val log = HttpLoggingInterceptor()
        log.setLevel(HttpLoggingInterceptor.Level.BASIC)
        return log
    }

    private val retrofit = Retrofit.Builder()
            .client(okHttpClient)
            .baseUrl(CovidApi.BASE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()

    private val covidApi: CovidApi = retrofit.create(CovidApi::class.java)


    val overAllHistory: Call<Result<OverAll>> = covidApi.overAllHistory

    val overAllLatest: Call<Result<OverAll>> = covidApi.overAllLatest

    val areaDataLatest: Call<Result<Area>> = covidApi.areaDataLatest

    val areaDataHistory: Call<Result<Area>> = covidApi.areaDataHistory
}